### PR TITLE
Add nanvl Spark function

### PR DESCRIPTION
**Pull request by zhli1142015**
Spark doc: https://spark.apache.org/docs/latest/api/sql/index.html#nanvl
Spark implementation: https://github.com/apache/spark/blob/443a957920737143834888726f067bc54acb97f2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala#L283

### CODE REVIEW

**Comment by zhli1142015** [comment]
@rui-mo , could you help review?
Thanks.

File: velox/functions/sparksql/tests/ArithmeticTest.cpp [hunk 1]
```diff
@@ -417,6 +417,34 @@ TEST_F(ArithmeticTest, isNanDouble) {
...
 
+TEST_F(ArithmeticTest, nanvlFloat) {
+  const auto nanvl = [&](std::optional<float> a, std::optional<float> b) {
+    return evaluateOnce<float>("nanvl(c0, c1)", a, b);
+  };
+
+  EXPECT_EQ(0.0f, nanvl(0.0f, kNan));
+  EXPECT_EQ(0.0f, nanvl(kNan, 0.0f));
+  EXPECT_EQ(0.0f, nanvl(0.0f / 0.0f, 0.0f));
+  EXPECT_EQ(std::nullopt, nanvl(0.0f / 0.0f, std::nullopt));
```
Comment (rui-mo) [inline velox/functions/sparksql/tests/ArithmeticTest.cpp:428]: Can we add test for the first value not being NaN, and the second value being null? That could help verify having nulls in one of the inputs does not imply null in the output.

File: velox/docs/functions/spark/math.rst [hunk 2]
```diff
@@ -139,6 +139,11 @@ Mathematical Functions
         SELECT CAST(1 as DECIMAL(20, 3)) * CAST(0 as DECIMAL(20, 3)); -- decimal 0.000000
         SELECT CAST(201e-38 as DECIMAL(38, 38)) * CAST(301e-38 as DECIMAL(38, 38)); -- decimal 0.0000000000000000000000000000000000000
 
+.. spark:function:: nanvl(x, y) -> [same as x]
+
+    Returns x if it's not NaN, or y otherwise. The types of x and y must be the same.
```
Comment (rui-mo) [inline velox/docs/functions/spark/math.rst:144]: Use \``x\`` and \``y\`` for variables.

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: velox/functions/sparksql/tests/ArithmeticTest.cpp (hunk 1 above)
Comment (zhli1142015) [reply velox/functions/sparksql/tests/ArithmeticTest.cpp:429]: Added, thanks.

File: velox/docs/functions/spark/math.rst [hunk 3]
```diff
@@ -139,6 +139,11 @@ Mathematical Functions
         SELECT CAST(1 as DECIMAL(20, 3)) * CAST(0 as DECIMAL(20, 3)); -- decimal 0.000000
         SELECT CAST(201e-38 as DECIMAL(38, 38)) * CAST(301e-38 as DECIMAL(38, 38)); -- decimal 0.0000000000000000000000000000000000000
 
+.. spark:function:: nanvl(x, y) -> [same as x]
+
+    Returns ``x`` if it's not NaN, or ``y`` otherwise. The types of ``x`` and ``y`` must be the same.
```
Comment (jinchengchenghh) [inline velox/docs/functions/spark/math.rst:144]: Looks like the x and y can be different datatype.
```
spark-sql (default)> select nanvl(cast(10.2 as double), cast(3.4 as float));
10.2
Time taken: 0.13 seconds, Fetched 1 row(s)
```

Comment (zhli1142015) [reply velox/docs/functions/spark/math.rst:144]: Spark has type coercion rule for this fucntion, so this fucntion should only get either (float, float) or (double, double). We don't need to handle different types inputs.
Thanks.
https://github.com/apache/spark/blob/3a8c1bca4fffaf4185bd0729f24f30e732ba72a4/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L497

Comment (jinchengchenghh) [reply velox/docs/functions/spark/math.rst:144]: ok

**Comment by zhli1142015** [comment]
Thanks @rui-mo .
@mbasmanova , could you help review?

File: velox/functions/sparksql/Arithmetic.h [hunk 4]
```diff
@@ -318,4 +318,31 @@ struct IsNanFunction {
...
+template <typename T>
+struct NanvlFunction {
+  template <typename TInput>
+  FOLLY_ALWAYS_INLINE void call(TInput& result, TInput a, TInput b) {
+    result = std::isnan(a) ? b : a;
+  }
+
+  template <typename TInput>
+  FOLLY_ALWAYS_INLINE bool
+  callNullable(TInput& result, const TInput* a, const TInput* b) {
```
Comment (mbasmanova) [inline velox/functions/sparksql/Arithmetic.h:331]: Please, update documentation to explain the behavior when x or y are NULL.

File: velox/functions/sparksql/RegisterArithmetic.cpp [hunk 5]
```diff
@@ -98,6 +98,10 @@ void registerArithmeticFunctions(const std::string& prefix) {
   VELOX_REGISTER_VECTOR_FUNCTION(udf_decimal_div, prefix + "divide");
   registerFunction<sparksql::IsNanFunction, bool, float>({prefix + "isnan"});
   registerFunction<sparksql::IsNanFunction, bool, double>({prefix + "isnan"});
+  registerFunction<sparksql::NanvlFunction, float, float, float>(
```
Comment (mbasmanova) [inline velox/functions/sparksql/RegisterArithmetic.cpp:101]: use registerBinaryFloatingPoint

**Review by mbasmanova — COMMENTED** [review]
Looks good % some comments.

**Comment by mbasmanova** [comment]
I wonder if nanvl should be implemented as a special form. It seems similar to coalesce which doesn't evaluate all arguments all the time. For example, what happens if x is not NaN, but y throws?

nanvl(1.2, c / 0.0)

I assume this return 1.2 without throwing.

**Review by jinchengchenghh — APPROVED** [review]

**Comment by zhli1142015** [comment]
> I wonder if nanvl should be implemented as a special form. It seems similar to coalesce which doesn't evaluate all arguments all the time. For example, what happens if x is not NaN, but y throws?
> 
> nanvl(1.2, c / 0.0)
> 
> I assume this returns 1.2 without throwing.

Thanks @mbasmanova  to poin this out. You are correct. The expected behavior is: argument y is only evaluated if necessary. 
I'm not sure if we should implement it as a special form. In my view, `nanvl(x, y)` and `if (isnan(x)) y else x` should have the same behavior. It is ok to perform this translation within the gluten codebase? @rui-mo  and @jinchengchenghh

**Comment by rui-mo** [comment]
Below shows the result I got in Spark for ANSI config being ON and OFF. It seems with special form, we can evaluate an expression only when needed, and exception will be not be thrown when its evaluation is skipped. This could help us align Velox's behavior with Spark when turning on ANSI config, and unneeded evaluation could be skipped for both.
But currently Gluten does not support ANSI ON mode, and this PR seems to give matched result with ANSI OFF mode. How do you think about it? @zhli1142015 @mbasmanova

ANSI OFF:
```
scala> spark.sql("select nanvl(1.2, 2 / 0.0)").show(false)
+---------------------+
|nanvl(1.2, (2 / 0.0))|
+---------------------+
|1.2                  |
+---------------------+


scala> spark.sql("select nanvl(2 / 0.0, 1.2)").show(false)
+---------------------+
|nanvl((2 / 0.0), 1.2)|
+---------------------+
|null                 |
+---------------------+
```

ANSI ON:
```
scala> spark.sql("select nanvl(1.2, 2 / 0.0)").show(false)
+---------------------+                                                         
|nanvl(1.2, (2 / 0.0))|
+---------------------+
|1.2                  |
+---------------------+


scala> spark.sql("select nanvl(2 / 0.0, 1.2)").show(false)
org.apache.spark.SparkArithmeticException: Division by zero.
```

**Comment by mbasmanova** [comment]
@rui-mo Rui, the testing above is not quite valid. When testing ansi off, make sure you use an expression that does throw as the second argument of nanvl.

**Comment by rui-mo** [comment]
@mbasmanova Thanks for the pointer. I understand.

**Comment by zhli1142015** [comment]
Hello @mbasmanova  and @rui-mo ,
Thank you for the review. I discussed it with Rui, and I believe that mapping the `nanvl` function to 'if + isnan' (in Gluten) is a simpler way, and it's behavior would be more aligned with Spark's. Therefore, I would like to close this PR. Please let me know if you have a different opinion.
Thanks.

**Comment by mbasmanova** [comment]
@zhli1142015 Thank you for letting me know. Makes sense.

**Pull request closed without merging**